securityscanutils/trivy_scanner.go

```diff
@@ -31,73 +41,71 @@ func RunTrivyScan(image, version, templateFile, output string) (bool, bool, erro
 	// operation. This timing is in the same ballpark as what k8s finds sensible
-	out, statusCode, err := executeTrivyScanWithRetries(
-		trivyScanArgs, 5,
-		func(attempt int) { time.Sleep(time.Duration((attempt^2)*2) * time.Second) },
-	)
+	scanCompleted, vulnerabilityFound, err := t.executeScanWithRetries(ctx, trivyScanArgs)
 
-	// Check if a vulnerability has been found
-	vulnFound := statusCode == VulnerabilityFoundStatusCode
-	// err will be non-nil if there is a non-zero status code
-	// so if the status code is the special "vulnerability found" status code,
-	// we don't want to report it as a regular error
-	if !vulnFound && err != nil {
-		// delete empty trivy output file that may have been created
+	if !scanCompleted {
+		// delete the empty trivy output file that may have been created
 		_ = os.Remove(output)
-		// swallow error if image is not found error, so that we can continue scanning releases
-		// even if some releases failed and we didn't publish images for those releases
-		// this error used to happen if a release was a pre-release and therefore images
-		// weren't pushed to the container registry.
-		// we have since filtered out non-release images from being scanned so this warning
-		// shouldn't occur, but leaving here in case there was another edge case we missed
-		if IsImageNotFoundErr(string(out)) {
-			log.Warnf("image %s not found for version %s", image, version)
-			return false, false, nil
-		}
-		return false, false, eris.Wrapf(err, "error running trivy scan on image %s, version %s, Logs: \n%s", image, version, string(out))
 	}
-	return true, vulnFound, nil
+
+	return scanCompleted, vulnerabilityFound, err
 }
 
-func executeTrivyScanWithRetries(trivyScanArgs []string, retryCount int,
-	backoffStrategy func(int)) ([]byte, int, error) {
-	if retryCount == 0 {
-		retryCount = 5
-	}
-	if backoffStrategy == nil {
-		backoffStrategy = func(attempt int) {
-			time.Sleep(time.Second)
-		}
-	}
+// executeScanWithRetries executes a trivy command (with retries and backoff)
+// and returns a tuple of (scanCompleted, vulnerabilitiesFound, error)
+func (t *TrivyScanner) executeScanWithRetries(ctx context.Context, scanArgs []string) (bool, bool, error) {
+	logger := contextutils.LoggerFrom(ctx)
 
 	var (
 		out        []byte
 		statusCode int
 		err        error
 	)
 
-	for attempt := 0; attempt < retryCount; attempt++ {
-		trivyScanCmd := exec.Command("trivy", trivyScanArgs...)
-		out, statusCode, err = executils.CombinedOutputWithStatus(trivyScanCmd)
+	for attempt := 0; attempt < t.scanMaxRetries; attempt++ {
+		trivyScanCmd := exec.Command("trivy", scanArgs...)
+		attemptStart := time.Now()
+		if t.executeCommand == nil {
+			return false, false, eris.Errorf("TrivyScanner has no executeCommand function defined")
+		}
+		out, statusCode, err = t.executeCommand(trivyScanCmd)
+		logger.Debugf("Trivy returned %d after %s", statusCode, time.Since(attemptStart).String())
```

Comment:
I like this log line! I wonder, now that this will occur many times, whether the other log messages should be at the info level?

Is there any concern with flooding the logs?

Reply:
Cleaned up the log statements to prevent flooding